### DISCUSSION TITLE
Where do I run `composer require` commands?

### DISCUSSION THREAD

**Discussion by jonathonbyrdziak** [alexstandiford/underpin#42] [Q&A] [answered]
It wasn't initially obvious that I needed to run `composer require underpin/underpin` in the root directory of Wordpress. I had to run this in an empty folder to see what the structure was before realizing where it needed to be ran.

And now that I'm through that hurdle, I still cannot figure out where to run the individual loaders like `https://github.com/Underpin-WP/admin-page-loader`.

To increase the adoption rate of underpin, it would seem that providing a clear walkthrough of each use case would help.

In this walkthrough it would be good to see what the suggested file structure should be.
As well as, seeing where you are installing each of the packages.
In the end, make your example available in a git repo so we can snag code snippets that you provided in the walkthrough.

Thanks so much. I'm looking for a rapid development framework to make Wordpress a bit more bearable. I'm hoping that this is it, but struggling to reverse engineer the assumptions made.

**Comment by alexstandiford** [comment, marked as answer]
Hey! Thanks for writing in! You're right - Underpin would benefit from having some example plugins. I'm actually working on a series of plugins for WP Dev Academy now that will provide some open-source examples, but really Underpin itself should have a basic example.

For now, the only example I have is the example used to set up BerlinDB and some custom tables here: https://github.com/Underpin-WP/berlindb-example

That should help you a-lot.

100% not required, but I also built a course that walks through building a plugin using underpin [here](https://www.wpdev.academy/course/beer-lister-plugin/).

Finally, if you haven't already, check out the [plugin boilerplate](https://github.com/Underpin-WP/underpin-plugin-boilerplate). You could potentially fork it and make your own version that embeds Underpin inside your plugin. By design, Underpin can be used as a package inside a WordPress plugin, or as a mu-plugin. The default behavior is to make it a mu-plugin because this framework is often used on custom WordPress websites, and multiple plugins and themes will want to utilize Underpin directly.

> in the root directory of Wordpress. I had to run this in an empty folder to see what the structure was before realizing where it needed to be ran.

Yep, that's because it's set up as a mu-plugin. You _can_ override where this is installed from though. If you want it to work directly inside of a WordPress plugin you need to add this to your `composer.json` file:

```json
  "extra": {
    "installer-paths": {
      "vendor/{$vendor}/{$name}": [
        "type:wordpress-muplugin"
      ]
    }
  },
```

This will tell composer to install any mu plugins as a regular ol' package inside the `vendor` directory. Once that's set up you can load in each loader using `composer require underpin/${loader-name}` where `${loader-name}` is the actual loader itself.

I think a good PR for this issue would be to update the `readme.md` file to include the note above about the `composer.json` file, and probably link to [Compoer's documentation on how this works](https://getcomposer.org/doc/faqs/how-do-i-install-a-package-to-a-custom-path-for-my-framework.md#how-do-i-install-a-package-to-a-custom-path-for-my-framework-) as well.

**Comment by alexstandiford** [comment]
Another issue that has crossed my mind - all of the loaders and extensions for Underpin should probably be labeled as a `wordpress-muplugin`, that way everything installs in the same directory. Right now if you install from the WP root directory you'll get Underpin in the `mu-plugins` dir, but everything else ends up in `vendor` on the root. Definitely not what you'd want.

**Comment by alexstandiford** [comment]
@jonathonbyrdziak I created a new boilerplate to make setting up Undeprin as a mu-plugin easier. [Check this out](https://github.com/Underpin-WP/mu-plugin-boilerplate)

This should give you the guidance needed to get moving.
